avocado/plugins/spawners/lxc.py

```diff
@@ -0,0 +1,221 @@
+        )
+
+        help_msg = "Release of the LXC container (depends on the choice of distro)"
+        settings.register_option(
+            section=section, key="release", help_msg=help_msg, default="32"
+        )
+
+        help_msg = "Architecture of the LXC container"
+        settings.register_option(
+            section=section, key="arch", help_msg=help_msg, default="i386"
+        )
+
+
+class LXCSpawner(Spawner, SpawnerMixin):
+
+    description = "LXC (container) based spawner"
+    METHODS = [SpawnMethod.STANDALONE_EXECUTABLE]
+
+    @staticmethod
+    def run_container_cmd(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            exitcode = container.attach_wait(
+                lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err
+            )
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    async def run_container_cmd_async(container, command):
+        with LXCStreamsFile() as tmp_out, LXCStreamsFile() as tmp_err:
+            pid = container.attach(
+                lxc.attach_run_command, command, stdout=tmp_out, stderr=tmp_err
+            )
+            loop = asyncio.get_event_loop()
+            _, exitcode = await loop.run_in_executor(
+                None, os.waitpid, pid, os.WUNTRACED
+            )
+            return exitcode, tmp_out.read(), tmp_err.read()
+
+    @staticmethod
+    def is_task_alive(runtime_task):
+        if runtime_task.spawner_handle is None:
+            return False
+
+        container = lxc.Container(runtime_task.spawner_handle)
+        if not container.defined:
+            LOG.debug(f"Container {runtime_task.spawner_handle} is not defined")
+            return False
+        if not container.running:
+            LOG.debug(
+                f"Container {runtime_task.spawner_handle} state is "
+                f"{container.state} instead of RUNNING"
+            )
+            return False
+
+        status, _, _ = LXCSpawner.run_container_cmd(
+            container, ["pgrep", "-r", "R,S", "avocado-runner"]
+        )
+        return status == 0
+
+    async def spawn_task(self, runtime_task):
+        self.create_task_output_dir(runtime_task)
+        task = runtime_task.task
+        # using the "python" symlink will result in the container default python version
+        entry_point_args = ["python", "-m", "avocado.core.nrunner", "task-run"]
+        entry_point_args.extend(task.get_command_args())
+
+        distro = self.config.get("spawner.lxc.distro")
+        release = self.config.get("spawner.lxc.release")
+        arch = self.config.get("spawner.lxc.arch")
+        create_hook = self.config.get("spawner.lxc.create_hook")
+
+        if not LXC_AVAILABLE:
+            msg = "LXC python bindings not available on the system"
+            runtime_task.status = msg
+            return False
+
+        container_id = self.cid
```

Comment:
@pevogam I know that there have been discussion about this already, but because the conversation is almost one year old and many tings had changed since then, I want to ask if this is still an actual problem. You said that, you have been using the lxc spawner for some time now, so I expect this was solved. But I am not understand how you reuse the containers since this variable is not defined anywhere. Can you give me more info, thanks?

Reply:
Indeed, this is one of the final remaining points and I have recorded all progress on it in this [comment](https://github.com/avocado-framework/avocado/pull/4158#discussion_r890722733). I tagged you there too in case you have a suggestion what approach would be most compatible with avocado's current approach and way of doing this.

Reply:
Let me elaborate on this here so that I see if you have any suggestion how to go about it. We have been using the LXC spawner for years but as a patch based 1:1 on the pull request here. This means that the `self.cid` is a trick we used in order to be able to set the container ID from code like:
```python
# TODO: move cid in constructor in the upstream PR
spawner = SpawnerDispatcher(job.config, job)[spawner_name].obj
spawner.cid = env_id
```
I admit this was a dirty and hackish way to achieve the purpose of setting this attribute that was there because we (and any other plugin) doesn't have direct access to the constructor of the spawner and instead accesses it via the spawner dispatcher.

To illustrate why the job config doesn't seem to be an option, I just applied the changes
```diff
-# TODO: move cid in constructor in the upstream PR
+job.config["spawner.lxc.name"] = env_id
 self.spawner = SpawnerDispatcher(job.config, job)[spawner_name].obj
-self.spawner.cid = env_id
```
on the side of caller (our plugin) and the following changes in the code here:
```diff
@@ -42,10 +42,15 @@ class LXCSpawnerInit(Init):
     description = "LXC (container) based spawner initialization"
 
     def initialize(self):
         section = "spawner.lxc"
 
+        help_msg = "Name for the LXC container"
+        settings.register_option(
+            section=section, key="cid", help_msg=help_msg, default="c1"
+        )
+
         help_msg = "Distro for the LXC container"
         settings.register_option(
             section=section, key="distro", help_msg=help_msg, default="fedora"
         )
 
@@ -118,21 +123,21 @@ class LXCSpawner(Spawner, SpawnerMixin):
         full_module_name = full_nonbogus_name = "avocado_vt.plugins.vt_runner"
         # using the "python" symlink will result in the container default python version
         entry_point_args = ["python", "-m", full_module_name, "task-run"]
         entry_point_args.extend(task.get_command_args())
 
+        container_id = self.config.get("spawner.lxc.cid")
         distro = self.config.get("spawner.lxc.distro")
         release = self.config.get("spawner.lxc.release")
         arch = self.config.get("spawner.lxc.arch")
         create_hook = self.config.get("spawner.lxc.create_hook")
 
         if not LXC_AVAILABLE:
             msg = "LXC python bindings not available on the system"
             runtime_task.status = msg
             return False
 
-        container_id = self.cid
         container = lxc.Container(container_id)
         if not container.defined:
             # Create the container rootfs
             if not container.create(
                 "download",
 ```
 but this will not work and inspecting the `self.config` shows the default value `c1` was used there instead. I admit I don't understand how the job and suite configs are parsed from the dispatcher but it seems to me that the settings are generated earlier and I cannot use such config to pass any information to the spawner. Using containers with different names or distros that are decided during runtime thus seems impossible using the current means.

Do you have any sane suggestion regarding this as somebody who is much more aware of the job and suite configs and how they work? How about some static variable on the class that we can set? I used `self.cid` before because I have seen similar other `self.x` attributes being available throughout the class and not being set inside its constructor and likely being set from the dispatcher too.

Reply:
```diff
-# TODO: move cid in constructor in the upstream PR
+job.config["spawner.lxc.name"] = env_id
 self.spawner = SpawnerDispatcher(job.config, job)[spawner_name].obj
-self.spawner.cid = env_id
```
Hi @pevogam this is definitely a possibility of, how you can pass value to the spawner, and it should work, but will be only persistent in spawner object which you created by Dispatcher. IMO, this is almost the same solution as the previous one with `self.spawner.cid = env_id`. And changing the avocado config during runtime is quite problematic, because IMO the configuration should static and based on parameters from user.

Unfortunately, it is difficult for me to help you with this, because I can not see the whole picture around the `cid`. I have a couple of questions about this:
* where is this code come form, is it from `runner_nrunner` plugin? (it is not part of this PR):
```
# TODO: move cid in constructor in the upstream PR
spawner = SpawnerDispatcher(job.config, job)[spawner_name].obj
spawner.cid = env_id
```
* How do you compute `env_id` during runtime? I think it should be part of this PR as well.
* On which layer do you want to reuse the containers? Inside suite, job, or even between avocado runs?

Because I don't have answers to those questions, I am not sure If my solution is accurate but. What about to store the container_id into avocado cache. We could create a file in `~/avocado/data/cahce/lxc_id` and get the id from it. Also, if the id has to be created before spawner, then I would propose to use `JobPreTests` and `JobPostTests` plugins to create the id and manipulate with avocado cache. With this solution, we can reuse the container for the entire suite without any changes to the `avocado.core` or `runner_nrunner.py`.

Reply:
> Hi @pevogam this is definitely a possibility of, how you can pass value to the spawner, and it should work, but will be only persistent in spawner object which you created by Dispatcher. IMO, this is almost the same solution as the previous one with `self.spawner.cid = env_id`. And changing the avocado config during runtime is quite problematic, because IMO the configuration should static and based on parameters from user.

Indeed, I surmised the same and upon testing I can confirm that the setting does not take effect like this and this diff doesn't really work.

> Unfortunately, it is difficult for me to help you with this, because I can not see the whole picture around the `cid`. I have a couple of questions about this:

Great, let me try to answer all of your questions so that we can see what is the best possible approach within avocado. In short: The main point is that the state machine and tasks can be spawned on multiple LXC containers (just like multiple Podman containers) each one of which can have its own separate name and is not "ephemeral", i.e. it can pre-exist before the spawned task and after the spawned task. The LXC containers are more complete subsystems in comparison to the podman containers whose life would typically begin and end with the task they are meant to isolate. Because of this, the state machine has to be aware of the LXC container "slots" (names of available containers to use) in advance and spawn a task within a free container among those (the environment of each container is assumed to be identical). To do this, we have to do some decision making first (not implemented in the task state machine here, perhaps we can simply assume some hard-coded name for this pull request to make the spawner code available and then discuss / propose PR for the way to invoke it) and eventually pass the cid or container name to the spawner so that we can spawn the task in a container of specific choice.

>     * where is this code come form, is it from `runner_nrunner` plugin? (it is not part of this PR):

In my case, our Avocado I2N plugin has its own runner which traverses a graph of dependencies and can use either of two spawners (process or LXC) to spawn a test/task. But you are right that we should make sure to provide proper conceptualization and bridging to use the LXC spawner with reusable container names here within the Avocado state machine. The main point is to make it possible to reuse free LXC container slots.

> ```
> # TODO: move cid in constructor in the upstream PR
> spawner = SpawnerDispatcher(job.config, job)[spawner_name].obj
> spawner.cid = env_id
> ```
> 
>     * How do you compute `env_id` during runtime? I think it should be part of this PR as well.

Exactly as above - considering a free container slot.

>     * On which layer do you want to reuse the containers? Inside suite, job, or even between avocado runs?

Even between avocado runs, as LXC containers are much more complete subsystems their setup could be rather more elaborate and must therefore be reused as much as possible.

> Because I don't have answers to those questions, I am not sure If my solution is accurate but. What about to store the container_id into avocado cache. We could create a file in `~/avocado/data/cahce/lxc_id` and get the id from it. Also, if the id has to be created before spawner, then I would propose to use `JobPreTests` and `JobPostTests` plugins to create the id and manipulate with avocado cache. With this solution, we can reuse the container for the entire suite without any changes to the `avocado.core` or `runner_nrunner.py`.

Sounds interesting, is the `JobPreTests` meant to be run before each test / task? Where should such a file creating/handling code be placed - inside the `JobPreTests`? Do you have some sample diff?

Reply:
@luckyh This is how I can imagine bringing parallelism to Avocado VT even without our company's custom scheduler runner: If this LXC spawner is used by Avocado VT instead of the process spawner and we provide a mechanism for Avocado to detect free LXC slots (like the pre-test logic above) and spawn new VT tests within them, then Avocado VT can simply use the default state machine and task scheduling provided by Avocado as always, changing only the spawner config setting and lifting away the `max-one-test-at-a-time` limit placed in its code.

Reply:
Also just to illustrate the above problem from a slightly different perspective: if we have a `RemoteSpawner` that would manage a cluster of physical machines and run tests in distributed fashion, we encounter the same problem - the scheduler-runner has to be aware of which physical machines have their tests completed and are free to run further tests in which case the spawner has to be provided with a runtime variable hostname (above container_id) in order to spawn the task at the correct remote address.

Reply:
> > Unfortunately, it is difficult for me to help you with this, because I can not see the whole picture around the `cid`. I have a couple of questions about this:
> 
> Great, let me try to answer all of your questions so that we can see what is the best possible approach within avocado. In short: The main point is that the state machine and tasks can be spawned on multiple LXC containers (just like multiple Podman containers) each one of which can have its own separate name and is not "ephemeral", i.e. it can pre-exist before the spawned task and after the spawned task. The LXC containers are more complete subsystems in comparison to the podman containers whose life would typically begin and end with the task they are meant to isolate. Because of this, the state machine has to be aware of the LXC container "slots" (names of available containers to use) in advance and spawn a task within a free container among those (the environment of each container is assumed to be identical). To do this, we have to do some decision making first (not implemented in the task state machine here, perhaps we can simply assume some hard-coded name for this pull request to make the spawner code available and then discuss / propose PR for the way to invoke it) and eventually pass the cid or container name to the spawner so that we can spawn the task in a container of specific choice.
> 
> > ```
> > * where is this code come form, is it from `runner_nrunner` plugin? (it is not part of this PR):
> > ```
> 
> In my case, our Avocado I2N plugin has its own runner which traverses a graph of dependencies and can use either of two spawners (process or LXC) to spawn a test/task. But you are right that we should make sure to provide proper conceptualization and bridging to use the LXC spawner with reusable container names here within the Avocado state machine. The main point is to make it possible to reuse free LXC container slots.
> 
@pevogam thank for the explanation. Now I think I have a much better understanding of what you are trying to achieve.

> > Because I don't have answers to those questions, I am not sure If my solution is accurate but. What about to store the container_id into avocado cache. We could create a file in `~/avocado/data/cahce/lxc_id` and get the id from it. Also, if the id has to be created before spawner, then I would propose to use `JobPreTests` and `JobPostTests` plugins to create the id and manipulate with avocado cache. With this solution, we can reuse the container for the entire suite without any changes to the `avocado.core` or `runner_nrunner.py`.
> 
> Sounds interesting, is the `JobPreTests` meant to be run before each test / task? At least this doesn't seem to be the case from its docstring:

Yes, the `JobPreTests` is meant to run before all tests. Therefore, is not suitable for your use-case.

> 
> At the same time this one describes actions taken _within_ the environment
> 
> ```
> class PreTest(Plugin):
>     """Base plugin interface for adding tasks before a test run.
> 
>     This interface helps to create `avocado.core.nrunner.Task` for running
>     additional actions inside spawner environment before Test.
>     """
> ```
Yes, the `PreTest` runs before each test, but inside the test environment. (For podman inside container, for process inside machine environment) IMO this is also not suitable for your use-case.

> 
> Where should such a file creating/handling code be placed then? Do you have some sample diff?

IMO, we can use very similar mechanism as we used with podman and requirement cache [here](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/spawners/podman.py#L370). In podman spawner we reuse images where the requirements are installed and I think it is a very similar problem to this, only instead of images it will be containers. IMO, everything can be done inside LXC spawner with a process like this:
1. LXC spawner will search for free container ID in avocado cache  (requirements cache can be used or some dedicated file). 
2. If cache doesn't have a free container ID, LXC spawner will create a new container.
3. LXC spawner will spawn a task in the container and will store the container ID into cache as occupied.
4. When the task finishes, the LXC spawner will mark the container ID in the cache as free.

Because the avocado cache is persistent between avocado runs, the creation of containers will be necessary only at the beginning of firs avocado run and can be reused until they are removed and avocado cache cleaned. We can even add avocado config option for user to provide list of container IDs to let LXC spawner use containers created by user before avocado run and store them into the cache. What do you think?

Reply:
Overall a very interesting suggestion. I have some more comments though:

> > Where should such a file creating/handling code be placed then? Do you have some sample diff?
> 
> IMO, we can use very similar mechanism as we used with podman and requirement cache [here](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/spawners/podman.py#L370). In podman spawner we reuse images where the requirements are installed and I think it is a very similar problem to this, only instead of images it will be containers. IMO, everything can be done inside LXC spawner with a process like this:
> 
>     1. LXC spawner will search for free container ID in avocado cache  (requirements cache can be used or some dedicated file).

How about something even simpler like a class attribute holding the cache in memory? After all we are talking about a simple list of a few short strings as container ids. I am not very familiar with the way the requirements are implemented in avocado, their format and types, mostly because in Avocado I2N we use our own scheduler (for many many years now). We could add such completion at the very least to make the pull request mergeable and then perhaps you could propose a PR that integrates this further into the task requirements paradigm? With my direct participation there we could make sure it works from all angles and is extendable enough.

>     2. If cache doesn't have a free container ID, LXC spawner will create a new container.

This might be a dangerous premise or should be configurable at the very least because new containers could consume a lot more RAM and disk memory.

>     3. LXC spawner will spawn a task in the container and will store the container ID into cache as occupied.
> 
>     4. When the task finishes, the LXC spawner will mark the container ID in the cache as free.

This could e.g. be easily done with a simple boolean dictionary within the LXCSpawner class where a set of predefined ID-s as keys could all have values True/False for free/occupied. But I wonder about something else: our Cartesian graph scheduler (just like avocado's state machine) might prefer to decide in which slot to run due to various prioritization and criteria. So our plugin generally needs to specify the container ID in advance and simply delegate the "spawning part" to the spawner. This then brings to two additional questions:

1) Could we also have some runtime task attribute which would indicate on which container the task is expected to be run? Similar cache handling like above then doesn't need to store anything in advance and can always consult with the cache if the required cid (container ID) is free or currently occupied (key created and flag set to true and not yet unset by a previous task). This would allow for further customization of how spawner(s) are used.

2) I find the general approach of putting the entire requirements logic a bit too mixed with the spawners in general also for `podman` and `process`. Shouldn't the spawner code just focus on starting a task within a given environment? I would have expected that the task requirements and other related dependencies are analyzed and handled by the state machine which then simply says "start this task in this environment" and this is what a spawner is responsible for. It seems the current code follows this approach more or less but something like free containers to run on sounds more scheduling-related than spawner related to me. But this is a minor thing, perhaps we should simply make sure that spawners could be used by more scheduling implementations.

> 
> Because the avocado cache is persistent between avocado runs, the creation of containers will be necessary only at the beginning of firs avocado run and can be reused until they are removed and avocado cache cleaned. We can even add avocado config option for user to provide list of container IDs to let LXC spawner use containers created by user before avocado run and store them into the cache. What do you think?

Indeed a file storage could be more persistent than storing in RAM via a class attribute. I was even thinking something along the lines of "requiring a container to be free could somehow be phrased as a task requirement under the current paradigm" but from what I see the `cache` object that podman uses is SQLite that expects certain concepts like "environments" (or podman images) and "requirements" to be provided and the "free container" to be reformulated as these, something that seems more appropriate for a follow up pull request.

What are your thoughts about all of this? I think you have a very good suggestion, I am just looking for a way to bring it down to smaller more easily digestable steps and keep compatibility with other ways to use the spawners like in our scheduler.

Reply:
> Overall a very interesting suggestion. I have some more comments though:
> 
> > > Where should such a file creating/handling code be placed then? Do you have some sample diff?
> > 
> > 
> > IMO, we can use very similar mechanism as we used with podman and requirement cache [here](https://github.com/avocado-framework/avocado/blob/master/avocado/plugins/spawners/podman.py#L370). In podman spawner we reuse images where the requirements are installed and I think it is a very similar problem to this, only instead of images it will be containers. IMO, everything can be done inside LXC spawner with a process like this:
> > ```
> > 1. LXC spawner will search for free container ID in avocado cache  (requirements cache can be used or some dedicated file).
> > ```
> 
> How about something even simpler like a class attribute holding the cache in memory? After all we are talking about a simple list of a few short strings as container ids. I am not very familiar with the way the requirements are implemented in avocado, their format and types, mostly because in Avocado I2N we use our own scheduler (for many many years now). We could add such completion at the very least to make the pull request mergeable and then perhaps you could propose a PR that integrates this further into the task requirements paradigm? With my direct participation there we could make sure it works from all angles and is extendable enough.

I agree that we can do something simpler than requirements cache for now and add requirements paradigm later.

> 
> > ```
> > 2. If cache doesn't have a free container ID, LXC spawner will create a new container.
> > ```
> 
> This might be a dangerous premise or should be configurable at the very least because new containers could consume a lot more RAM and disk memory.

Sure, we can add config value to set maximum number of containers. 
> 
> > ```
> > 3. LXC spawner will spawn a task in the container and will store the container ID into cache as occupied.
> > 
> > 4. When the task finishes, the LXC spawner will mark the container ID in the cache as free.
> > ```
> 
> This could e.g. be easily done with a simple boolean dictionary within the LXCSpawner class where a set of predefined ID-s as keys could all have values True/False for free/occupied. But I wonder about something else: our Cartesian graph scheduler (just like avocado's state machine) might prefer to decide in which slot to run due to various prioritization and criteria. So our plugin generally needs to specify the container ID in advance and simply delegate the "spawning part" to the spawner. This then brings to two additional questions:
> 
>     1. Could we also have some runtime task attribute which would indicate on which container the task is expected to be run? Similar cache handling like above then doesn't need to store anything in advance and can always consult with the cache if the required cid (container ID) is free or currently occupied (key created and flag set to true and not yet unset by a previous task). This would allow for further customization of how spawner(s) are used.

Yes, for this can be used `RuntimeTask.spawner_handle` variable. But IIUIC this solution needs changes to the state machine to schedule tasks based on cache and we don't have such support for now. But maybe you can use it in your Cartesian graph scheduler.

> 
>     2. I find the general approach of putting the entire requirements logic a bit too mixed with the spawners in general also for `podman` and `process`. Shouldn't the spawner code just focus on starting a task within a given environment? I would have expected that the task requirements and other related dependencies are analyzed and handled by the state machine which then simply says "start this task in this environment" and this is what a spawner is responsible for. It seems the current code follows this approach more or less but something like free containers to run on sounds more scheduling-related than spawner related to me. But this is a minor thing, perhaps we should simply make sure that spawners could be used by more scheduling implementations.

Yes, you are right that the task scheduling is a little bit mixed with spawners and we can definitely do improvements here, but that is for another PR. So I would suggest to stick to changes in spawner even with the scheduling so we can merge this.

> 
> 
> > Because the avocado cache is persistent between avocado runs, the creation of containers will be necessary only at the beginning of firs avocado run and can be reused until they are removed and avocado cache cleaned. We can even add avocado config option for user to provide list of container IDs to let LXC spawner use containers created by user before avocado run and store them into the cache. What do you think?
> 
> Indeed a file storage could be more persistent than storing in RAM via a class attribute. I was even thinking something along the lines of "requiring a container to be free could somehow be phrased as a task requirement under the current paradigm" but from what I see the `cache` object that podman uses is SQLite that expects certain concepts like "environments" (or podman images) and "requirements" to be provided and the "free container" to be reformulated as these, something that seems more appropriate for a follow up pull request.
> 
> What are your thoughts about all of this? I think you have a very good suggestion, I am just looking for a way to bring it down to smaller more easily digestable steps and keep compatibility with other ways to use the spawners like in our scheduler.

I think we are on the same page here. Unfortunately, I don't know how your scheduler works and if it is possible to make it compatible with avocado state machine, but for now I would do the cache handling inside LXC spawner. And after we merge this, we can discuss how to improve cache handling in avocado state machine.

Reply:
Alright, I pushed a possible implementation but with a caveat that we cannot simply guess a container name since in LXC the containers are not some UUID generated hashes but actual custom name. The typical naming we use is like c1, c2, c3, etc. which could be a good default but the LXC spawner cannot guess this for us or anyone else. So right now I think it is better if people simply define the containers they want to use via a `slots` config or a custom scheduler (likely also config for it).

I also added minimal isolation test coverage for this less-boundary-like functionality.